### PR TITLE
Initial support for gcs buckets

### PR DESCRIPTION
**Pull request by vinceatbluelabs**
This is enough code to get the following things to work:
* mvrec url2url gs://gs-bucket/foo.csv s3://s3-bucket/foo.csv 
* mvrec table2url redshift vbroz foo gs://gs-bucket/foo.csv
* mvrec table2recordsdir redshift vbroz foo gs://gs-bucket/mumble/
* mvrec url2url s3://s3-bucket/foo.csv gs://gcs-bucket/foo.csv

Notes:
* This PR doesn't add direct bulk export or import from BigQuery to GCS; the changes seemed significant enough to get moving without it.
* All copies between gs and other schemes is by copying bytes through memory in records-mover - this is not going to work for super large files.  Perhaps a cloud native solution can be found for that need in the future, and perhaps records-mover could tie into it somehow and use it automatically.

### CODE REVIEW

File: records_mover/session.py [hunk 1]
```diff
@@ -2,14 +2,15 @@
...
-from typing import Union, Optional, IO
+from typing import Union, Optional, IO, Dict
 from .url.resolver import UrlResolver
 from db_facts.db_facts_types import DBFacts
 from enum import Enum
 from records_mover.creds.creds_via_lastpass import CredsViaLastPass
 from records_mover.creds.creds_via_airflow import CredsViaAirflow
 from records_mover.creds.creds_via_env import CredsViaEnv
 from records_mover.logging import set_stream_logging
+from records_mover.utils import lazyprop
```
Comment (bluelabsbutler) [inline records_mover/session.py:13]: F401 'records_mover.utils.lazyprop' imported but unused

File: records_mover/session.py [hunk 2]
```diff
@@ -213,5 +249,7 @@ def set_stream_logging(self,
 
     @property
     def records(self) -> Records:
-        return Records(db_driver=self.db_driver,
-                       url_resolver=self.url_resolver)
+        if self._records is None:
+            self._records =  Records(db_driver=self.db_driver,
```
Comment (bluelabsbutler) [inline records_mover/session.py:253]: E222 multiple spaces after operator

File: records_mover/session.py [hunk 3]
```diff
@@ -109,15 +118,28 @@ def __init__(self,
...
         self._default_aws_creds_name = default_aws_creds_name
+        self._default_gcp_creds_name = default_gcp_creds_name
         self._scratch_s3_url = scratch_s3_url
         self.creds = creds
-        url_resolver_kwargs = {}
+        self._records: Optional[Records] = None
+
+    @property
+    def url_resolver(self) -> UrlResolver:
+        # TODO: Make cache
```
Comment (bluelabsbutler) [inline records_mover/session.py:133]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/url/s3/s3_file_url.py [hunk 4]
```diff
@@ -55,6 +55,7 @@ def open(self, mode: str = "rb") -> IO[bytes]:
                            key_id=self.key,
                            mode=mode,
                            session=self._boto3_session)
+        # TODO: upgrade probably broke this - look it up
```
Comment (bluelabsbutler) [inline records_mover/url/s3/s3_file_url.py:58]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/cli.py [hunk 5]
```diff
@@ -82,14 +88,18 @@ def main() -> None:
     # https://stackoverflow.com/questions/15405636/pythons-argparse-to-show-programs-version-with-prog-and-version-string-formatt
     parser.add_argument('-V', '--version', action='version', version="%(prog)s ("+__version__+")")
     subparsers = parser.add_subparsers(help='subcommand_help')
+    from records_mover import Session
+    bootstrap_session = Session()
```
Comment (vinceatbluelabs) [inline records_mover/records/cli.py:92]: We're doing more work now during Session initialization by fetching GCP credentials - which, combined with the annoying logs suppressed above, made doing this in a tight loop pretty apparent.  Instead, let's do this once and pass it in.

File: records_mover/session.py [hunk 6]
```diff
@@ -213,5 +248,7 @@ def set_stream_logging(self,
 
     @property
     def records(self) -> Records:
```
Comment (vinceatbluelabs) [inline records_mover/session.py:250]: Instead, this gets called in a loop, and we can easily stash this.

File: records_mover/url/gcs/gcs_directory_url.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        return GCSDirectoryUrl(url, gcs_client=self.client, gcp_credentials=self.credentials)
+
+    def directory_in_this_directory(self, directory_name: str) -> 'GCSDirectoryUrl':
+        return self._directory(f"{self.url}{directory_name}/")
+
+    def purge_directory(self) -> None:
+        if not self.is_directory():
+            raise ValueError("Not a directory")
+        # https://stackoverflow.com/questions/53165246/how-to-delete-gcs-folder-from-python
+        # TODO: see deprecation - https://googleapis.dev/python/storage/latest/buckets.html#google.cloud.storage.bucket.Bucket.list_blobs
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:34]: E501 line too long (137 > 100 characters)

Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:34]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/url/gcs/gcs_directory_url.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        blobs = self.bucket_obj.list_blobs(prefix=self.blob)
+        for blob in blobs:
+            # TODO: Does this work recursively?
+            blob.delete()
+
+    def files_and_directories_in_directory(self) -> List[Union[BaseFileUrl, BaseDirectoryUrl]]:
+        # https://stackoverflow.com/a/57099354/9795956
+
+        # TODO: Test this manully after creating a directory of a directory
+        file_urls = {}
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:44]: F841 local variable 'file_urls' is assigned to but never used

File: records_mover/url/gcs/gcs_directory_url.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        directory_urls = {}
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:45]: F841 local variable 'directory_urls' is assigned to but never used

File: records_mover/url/gcs/gcs_directory_url.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        service = googleapiclient.discovery.build('storage', 'v1',
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:46]: F841 local variable 'service' is assigned to but never used

File: records_mover/url/gcs/gcs_directory_url.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        return self._directory(f"{self.url}{directory_name}/")
+
+    def purge_directory(self) -> None:
+        if not self.is_directory():
+            raise ValueError("Not a directory")
+        # https://stackoverflow.com/questions/53165246/how-to-delete-gcs-folder-from-python
+        # TODO: see deprecation - https://googleapis.dev/python/storage/latest/buckets.html#google.cloud.storage.bucket.Bucket.list_blobs
+        blobs = self.bucket_obj.list_blobs(prefix=self.blob)
+        for blob in blobs:
+            # TODO: Does this work recursively?
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:37]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/url/gcs/gcs_directory_url.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            blob.delete()
+
+    def files_and_directories_in_directory(self) -> List[Union[BaseFileUrl, BaseDirectoryUrl]]:
+        # https://stackoverflow.com/a/57099354/9795956
+
+        # TODO: Test this manully after creating a directory of a directory
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:43]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/session.py [hunk 13]
```diff
@@ -2,8 +2,9 @@
 from .database import db_facts_from_env
 from .records.records import Records
 from .url.base import BaseFileUrl, BaseDirectoryUrl
-from typing import Union, Optional, IO
+from typing import Union, Optional, IO, Dict
 from .url.resolver import UrlResolver
+from .utils import lazyprop
```
Comment (bluelabsbutler) [inline records_mover/session.py:7]: F401 '.utils.lazyprop' imported but unused

File: records_mover/url/gcs/gcs_directory_url.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
+from urllib.parse import urlparse, unquote
+from records_mover.url import BaseDirectoryUrl, BaseFileUrl
+import google.auth.credentials
+import google.cloud.storage
+import googleapiclient.discovery
+from typing import List, Union
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:6]: F401 'typing.Union' imported but unused

File: records_mover/url/gcs/gcs_directory_url.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+        self.scheme = parsed.scheme
+        self.bucket = parsed.netloc
+        self.client = gcs_client
+        self.credentials = gcp_credentials
+        self.bucket_obj = self.client.bucket(self.bucket)
+
+    def _directory(self, url: str) -> 'GCSDirectoryUrl':
+        return GCSDirectoryUrl(url, gcs_client=self.client, gcp_credentials=self.credentials)
+
+    def _file(self, url: str) -> 'GCSFileUrl':
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:27]: F821 undefined name 'GCSFileUrl'

File: records_mover/session.py [hunk 16]
```diff
@@ -194,6 +221,42 @@ def _boto3_session(self) -> Optional['boto3.session.Session']:
         else:
             return self.creds.boto3_session(self._default_aws_creds_name)
 
+    def _gcs_creds(self) -> Optional['google.auth.credentials.Credentials']:
+        # TODO: Should save this
```
Comment (bluelabsbutler) [inline records_mover/session.py:225]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/url/gcs/gcs_file_url.py [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+            # google.api_core.exceptions.NotFound: 404 blob
+            # SkosHyDshbo/_manifest not found in bucket-name-here
+            #
+            # smart-open version: 2.0
+            raise FileNotFoundError(f"{self} not found")
+
+    def _directory(self, url: str) -> GCSDirectoryUrl:
+        return GCSDirectoryUrl(url, gcs_client=self.client, gcp_credentials=self.credentials)
+
+    def _blob_obj(self) -> 'Blob':
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_file_url.py:39]: F821 undefined name 'Blob'

File: tests/unit/test_cli_session.py [hunk 18]
```diff
@@ -1,5 +1,6 @@
 from records_mover import Session
 import unittest
+import google.auth
```
Comment (bluelabsbutler) [inline tests/unit/test_cli_session.py:3]: F401 'google.auth' imported but unused

File: records_mover/url/gcs/gcs_directory_url.py [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+        return GCSFileUrl(url, gcs_client=self.client, gcp_credentials=self.credentials)
+
+    def directory_in_this_directory(self, directory_name: str) -> 'GCSDirectoryUrl':
+        return self._directory(f"{self.url}{directory_name}/")
+
+    def purge_directory(self) -> None:
+        if not self.is_directory():
+            raise ValueError("Not a directory")
+        # https://stackoverflow.com/questions/53165246/how-to-delete-gcs-folder-from-python
+        # TODO: see deprecation -
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:40]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: types/stubs/google/cloud/storage/bucket/__init__.py [hunk 20]
```diff
@@ -0,0 +1,10 @@
+from typing import Iterator, Optional
```
Comment (bluelabsbutler) [inline types/stubs/google/cloud/storage/bucket/__init__.py:1]: F401 'typing.Iterator' imported but unused

Comment (bluelabsbutler) [inline types/stubs/google/cloud/storage/bucket/__init__.py:1]: F401 'typing.Optional' imported but unused

File: tests/unit/url/gcs/test_gcs_file_url.py [hunk 21]
```diff
@@ -0,0 +1,16 @@
+from records_mover.url.gcs.gcs_file_url import GCSFileUrl
+from mock import patch, Mock
```
Comment (bluelabsbutler) [inline tests/unit/url/gcs/test_gcs_file_url.py:2]: F401 'mock.patch' imported but unused

File: records_mover/url/gcs/gcs_directory_url.py [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+        for dir_loc in self.directories_in_directory():
+            dir_loc.purge_directory()
+        for file_loc in self.files_in_directory():
+            file_loc.delete()
+        self.bucket_obj.blob(self.blob).delete()
+
+    def files_in_directory(self) -> List[BaseFileUrl]:
+        prefix = self.blob
+        if prefix == '/':
+            # TODO: Does this ever come out as '/'?  If so, doe sit need to be stripped from others?
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:46]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/url/gcs/gcs_directory_url.py [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+        return [self._file(file_url) for file_url in file_urls]
+
+    def directories_in_directory(self) -> List[BaseDirectoryUrl]:
+        # https://stackoverflow.com/a/57099354/9795956
+
+        service = googleapiclient.discovery.build('storage', 'v1',
+                                                  credentials=self.credentials)
+        prefix = self.blob
+        if prefix == '/':
+            # TODO: Does this ever come out as '/'?  If so, doe sit need to be stripped from others?
```
Comment (bluelabsbutler) [inline records_mover/url/gcs/gcs_directory_url.py:71]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/url/base.py [hunk 24]
```diff
@@ -110,6 +130,7 @@ def string_contents(self, encoding: str = 'utf-8') -> str:
             return f.read().decode(encoding)
 
     def json_contents(self) -> Optional[JsonValue]:
+        "Return contents of this JSON file as a Python object, or None if file is empty (zero bytes)"
```
Comment (bluelabsbutler) [inline records_mover/url/base.py:133]: E501 line too long (101 > 100 characters)

File: tests/unit/records/test_records_format.py [hunk 25]
```diff
@@ -0,0 +1,24 @@
+import unittest
+from records_mover.records.records_format import RecordsFormat, DelimitedRecordsFormat
+from records_mover.url.base import BaseDirectoryUrl
```
Comment (bluelabsbutler) [inline tests/unit/records/test_records_format.py:3]: F401 'records_mover.url.base.BaseDirectoryUrl' imported but unused

File: tests/unit/records/test_records_format.py [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+from mock import Mock, patch, call
```
Comment (bluelabsbutler) [inline tests/unit/records/test_records_format.py:4]: F401 'mock.call' imported but unused

File: tests/unit/records/test_records_format.py [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+import json
```
Comment (bluelabsbutler) [inline tests/unit/records/test_records_format.py:5]: F401 'json' imported but unused

File: records_mover/session.py [hunk 28]
```diff
@@ -194,6 +217,44 @@ def _boto3_session(self) -> Optional['boto3.session.Session']:
...
+    def _gcs_creds(self) -> Optional['google.auth.credentials.Credentials']:
+        try:
+            import google.auth.exceptions
+            if self._default_gcp_creds_name is None:
+                import google.auth
+                credentials, project = google.auth.default()
+                return credentials
+            else:
+                return self.creds.gcs(self._default_gcp_creds_name)
+        except OSError:  # TODO: google.auth.exceptions.DefaultCredentialsError):
```
Comment (bluelabsbutler) [inline records_mover/session.py:229]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/session.py [hunk 29]
```diff
@@ -2,7 +2,7 @@
 from .database import db_facts_from_env
 from .records.records import Records
 from .url.base import BaseFileUrl, BaseDirectoryUrl
-from typing import Union, Optional, IO
+from typing import Union, Optional, IO, Dict
```
Comment (bluelabsbutler) [inline records_mover/session.py:5]: F401 'typing.Dict' imported but unused

File: records_mover/session.py [hunk 30]
```diff
@@ -190,9 +220,58 @@ def _boto3_session(self) -> Optional['boto3.session.Session']:
...
+            #     credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create
+            #     credentials and re-run the application. For more information, please see
+            #     https://cloud.google.com/docs/authentication/getting-started
+            logger.debug("google.cloud.storage not configured",
+                         exc_info=True)
+            self.__gcs_creds = None
+        return self.__gcs_creds
+
+    def _gcs_client(self) -> Optional['google.cloud.storage.Client']:
+        # TODO: cache
```
Comment (bluelabsbutler) [inline records_mover/session.py:255]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/url/__init__.py [hunk 31]
```diff
@@ -1,6 +1,7 @@
 from urllib.parse import urlparse
 from .base import BaseDirectoryUrl, BaseFileUrl
 import logging
+import inspect
```
Comment (bluelabsbutler) [inline records_mover/url/__init__.py:4]: F401 'inspect' imported but unused

File: records_mover/url/resolver.py [hunk 32]
```diff
@@ -1,14 +1,62 @@
...
+        init_urls()
+        parsed_url = urlparse(url)
+        if parsed_url.scheme in directory_url_ctors:
+            ctor = directory_url_ctors[parsed_url.scheme]
+            out = ctor(url, **self.kwargs_for_function(ctor))
+            if not isinstance(out, BaseDirectoryUrl):
+                raise TypeError(f"Not a directory url: {url}")
+            return out
+        else:
+            raise NotImplementedError(f"Teach me how to create DirectoryUrls for {parsed_url.scheme}")
```
Comment (bluelabsbutler) [inline records_mover/url/resolver.py:62]: E501 line too long (102 > 100 characters)

File: records_mover/url/resolver.py [hunk 33]
```diff
@@ -1,14 +1,62 @@
+from urllib.parse import urlparse
 from .base import BaseFileUrl, BaseDirectoryUrl
-from . import FileUrl, DirectoryUrl
-from typing import Any
+# TODO: Should I move these data structures here?
```
Comment (bluelabsbutler) [inline records_mover/url/resolver.py:3]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/url/resolver.py [hunk 34]
```diff
@@ -1,14 +1,66 @@
...
+            if not isinstance(out, BaseFileUrl):
+                raise TypeError(f"Not a file url: {url}")
+            return out
+        else:
+            raise NotImplementedError(f"Teach me how to create FileUrls for {parsed_url.scheme}")
+
+    def kwargs_for_function(self, fn: Callable) -> Dict[str, Any]:
+        parameters: Dict[str, Type] = inspect.signature(fn).parameters
+        out: Dict[str, Any] = {}
+        # TODO: Why shouldn't boto3_session_getter also be nullable
```
Comment (bluelabsbutler) [inline records_mover/url/resolver.py:46]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: .circleci/config.yml [hunk 35]
```diff
@@ -488,8 +488,8 @@ workflows:
           extras: '[redshift-binary,itest]'
           python_version: "3.6"
           db_name: demo-itest
-          requires:
-            - test-3.8
+#          requires:  # TODO
```
Comment (bluelabsbutler) [inline .circleci/config.yml:491]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/url/resolver.py [hunk 36]
```diff
@@ -1,14 +1,107 @@
+from urllib.parse import urlparse
 from .base import BaseFileUrl, BaseDirectoryUrl
-from . import FileUrl, DirectoryUrl
-from typing import Any
+import inspect
+from typing import Callable, Optional, Dict, Any, Type, Union, TYPE_CHECKING
+if TYPE_CHECKING:
+    import google.cloud.storage  # noqa
+    import boto3.session  # noqa
+import logging
```
Comment (bluelabsbutler) [inline records_mover/url/resolver.py:8]: E402 module level import not at top of file

File: records_mover/url/resolver.py [hunk 37]
```diff
@@ -1,14 +1,109 @@
...
+            file_url_ctors['s3'] = S3Url
+        if GCSFileUrl is not None:
+            file_url_ctors['gs'] = GCSFileUrl
+        file_url_ctors['file'] = FilesystemFileUrl
+
+        file_url_ctors['http'] = HttpFileUrl
+        file_url_ctors['https'] = HttpFileUrl
+
+
+# TODO: Instead of three functions, how about passing in an object that includes an interface?
```
Comment (bluelabsbutler) [inline records_mover/url/resolver.py:52]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/url/resolver.py [hunk 38]
```diff
@@ -1,14 +1,111 @@
...
+            if not isinstance(out, BaseFileUrl):
+                raise TypeError(f"Not a file url: {url}")
+            return out
+        else:
+            raise NotImplementedError(f"Teach me how to create FileUrls for {parsed_url.scheme}")
+
+    def kwargs_for_function(self, fn: Callable) -> Dict[str, Any]:
+        parameters: Dict[str, Type] = inspect.signature(fn).parameters
+        out: Dict[str, Any] = {}
+        # TODO: Can I get mypy to complain that I'm assigning a None
```
Comment (bluelabsbutler) [inline records_mover/url/resolver.py:88]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/url/resolver.py [hunk 39]
```diff
@@ -1,14 +1,127 @@
+from urllib.parse import urlparse
 from .base import BaseFileUrl, BaseDirectoryUrl
-from . import FileUrl, DirectoryUrl
-from typing import Any
+import inspect
+from typing import Callable, Optional, Dict, Any, Type, Union, TYPE_CHECKING
```
Comment (bluelabsbutler) [inline records_mover/url/resolver.py:4]: F401 'typing.Any' imported but unused

File: metrics/bigfiles_high_water_mark [hunk 40]
```diff
@@ -1 +1 @@
-968
+989
```
Comment (vinceatbluelabs) [inline metrics/bigfiles_high_water_mark:1]: `session.py` has made the list and is bumping this up.  While I do expect some refactoring of this to happen in future PRs (in particular moving some of the scratch bucket and default cred functionality into the Creds class hierarchy), I'm open to ideas on how to reorganize!

File: records_mover/airflow/hooks/records_hook.py [hunk 41]
```diff
@@ -28,7 +28,9 @@ def _get_boto3_session(self) -> boto3.session.Session:
 
     @property
     def _url_resolver(self) -> UrlResolver:
-        return UrlResolver(boto3_session=self._get_boto3_session())
+        return UrlResolver(boto3_session_getter=self._get_boto3_session,
```
Comment (vinceatbluelabs) [inline records_mover/airflow/hooks/records_hook.py:31]: Up until this PR, we had been eagerly fetching the Boto3 Session object.  That's been fine, as it's a pretty fast thing to fetch, usually.

Unfortunately, that is not true of GCP creds, at least in Application Default Credentials form.  There's a significant delay as they are returned, and an obnoxious "we've deprecated this functionality but have no realistic replacement" log message, to boot!

As a result, I put in machinery to both load this lazily only when we're about and cache it.  I applied the same machinery to the boto3 credentials for consistency's sake.

Much of this responsibility lives in the UrlResolver class, which is responsible for the caching and lazy loading.  The Session object (and this RecordsHook) are responsible for passing in a function that returns the creds (or None if none are configured) when called.

File: records_mover/airflow/hooks/records_hook.py [hunk 42]
```diff
@@ -28,7 +28,9 @@ def _get_boto3_session(self) -> boto3.session.Session:
...
+                           gcs_client_getter=lambda: None,
+                           gcp_credentials_getter=lambda: None)
```
Comment (vinceatbluelabs) [inline records_mover/airflow/hooks/records_hook.py:33]: I haven't converted this over to look up GCS GCP creds yet.  It seems like there's starting to be some overlap between RecordsHook and CredsViaAirflow.  

Could it be time to bring in the Session object here?  Or at least an instance of CredsViaAirflow?

File: records_mover/session.py [hunk 43]
```diff
@@ -77,6 +78,12 @@ def _infer_default_aws_creds_name(session_type: str) -> Optional[str]:
     return None
 
 
+def _infer_default_gcp_creds_name(session_type: str) -> Optional[str]:
+    if session_type == 'airflow':
+        return 'google_cloud_default'
+    return None
```
Comment (vinceatbluelabs) [inline records_mover/session.py:84]: I tried to keep the AWS and GCP creds fetching to be as similar as possible here.

File: records_mover/session.py [hunk 44]
```diff
@@ -99,17 +106,22 @@ def _infer_creds(session_type: str) -> BaseCreds:
...
+# https://github.com/python/typing/issues/236
 class PleaseInfer(Enum):
-    # This is a mypy-friendly way of doing a singleton object:
-    #
-    # https://github.com/python/typing/issues/236
+    token = 1
+
+
+class NotYetFetched(Enum):
     token = 1
```
Comment (vinceatbluelabs) [inline records_mover/session.py:117]: "None" for a cached cred means "no cred is available".  NotYetFetched means "we haven't checked yet".

File: records_mover/url/base.py [hunk 45]
```diff
@@ -1,4 +1,5 @@
 from contextlib import contextmanager
```
Comment (vinceatbluelabs) [inline records_mover/url/base.py:1]: There was a lot of common code from the S3 implementation that made sense to be generic here.

File: records_mover/url/base.py [hunk 46]
```diff
@@ -133,8 +155,9 @@ def download_fileobj(self, output_fileobj: IO[bytes]) -> None:
 
     def store_string(self, contents: str) -> None:
         "Save the specified string to the file."
-        with self.open(mode='wt') as f:
-            f.write(contents)
+
+        with self.open(mode='wb') as f:
+            f.write(contents.encode('utf-8'))
```
Comment (vinceatbluelabs) [inline records_mover/url/base.py:160]: The smartopen stuff only works with fileobjs opened in binary mode, so let's just do that.

File: records_mover/url/resolver.py [hunk 47]
```diff
@@ -1,14 +1,125 @@
+from urllib.parse import urlparse
```
Comment (vinceatbluelabs) [inline records_mover/url/resolver.py:1]: This module is now responsible for doing the lazy evaluation of the cloud creds.

File: records_mover/url/resolver.py [hunk 48]
```diff
@@ -1,14 +1,125 @@
...
+        directory_url_ctors['file'] = FilesystemDirectoryUrl
+    if len(file_url_ctors) == 0:
+        if S3Url is not None:
+            file_url_ctors['s3'] = S3Url
+        if GCSFileUrl is not None:
+            file_url_ctors['gs'] = GCSFileUrl
+        file_url_ctors['file'] = FilesystemFileUrl
+
+        file_url_ctors['http'] = HttpFileUrl
+        file_url_ctors['https'] = HttpFileUrl
```
Comment (vinceatbluelabs) [inline records_mover/url/resolver.py:50]: This logic seemed natural to roll in.

File: metrics/bigfiles_high_water_mark [hunk 49]
```diff
@@ -1 +1 @@
-968
+983
```
Comment (vinceatbluelabs) [inline metrics/bigfiles_high_water_mark:1]: session.py has made the list and is bumping this up. While I do expect some refactoring of this to happen in future PRs (in particular moving some of the scratch bucket and default cred functionality into the Creds class hierarchy), I'm open to ideas on how to reorganize!

**Comment by vinceatbluelabs** [comment]
Hi @cwegrzyn!  A review on this one would be super helpful in moving some other dependent items along.

**Pull request merged**